demo/app/src/App.js

```diff
@@ -0,0 +1,349 @@
+
+const ad = mainPage.createAd(el, ${JSON.stringify(
+      getAdConfiguration(),
+      null,
+      "  "
+    )});
+`;
+
+    setCodePreview(sourceCode);
+  };
+
+  useEffect(() => {
+    setSourceCodePreview();
+    // eslint-disable-next-line
+  }, [
+    path,
+    provider,
+    targeting,
+    refreshRate,
+    offset,
+    autoRefreshPlugin,
+    autoRenderPlugin,
+    debugPlugin,
+    loggingPlugin,
+    responsivePlugin,
+    stickyPlugin,
+    selectedDesktopSizes,
+    selectedTabletSizes,
+    selectedMobileSizes
+  ]);
+
+  const previewAd = () => {
+    if (window.previewInstance) {
+      window.previewInstance.destroy();
+    }
+
+    const el = document.getElementById("ad-slot-1");
+
+    const bucket = new window.AdJS.Bucket(
+      window.AdJS.Networks[provider],
+      getBucketConfiguration()
+    );
+
+    window.previewInstance = bucket.createAd(el, getAdConfiguration());
+  };
+
+  return (
+    <div className="config">
+      <StatefulTabs initialState={{ activeKey: "0" }}>
+        <Tab title="Settings">
+          <Block display="grid" gridTemplateColumns="130px 1fr">
+            <label>Network</label>
+            <Input
+              disabled
+              value={provider}
+              onChange={event => setProvider(event.target.value)}
+            />
+          </Block>
+
+          <Block display="grid" gridTemplateColumns="130px 1fr">
+            <label>Path</label>
+            <Input
+              value={path}
+              onChange={event => setPath(event.target.value)}
+            />
+          </Block>
+
+          <Block display="grid" gridTemplateColumns="130px 1fr">
+            <label>Targeting</label>
+            <Input
+              value={targeting}
+              onChange={event => setTargeting(event.target.value)}
+            />
+          </Block>
+
+          <Block display="grid" gridTemplateColumns="130px 1fr">
+            <label>Refresh</label>
+            <Slider
+              min={0}
+              max={60}
+              step={5}
+              value={[refreshRate]}
+              onChange={({ value }) => setRefreshRate(value[0])}
+              overrides={{
+                TickBar: {
+                  style: {
+                    display: "none"
+                  }
+                }
+              }}
+            />
+          </Block>
+
+          <Block display="grid" gridTemplateColumns="130px 1fr">
+            <label>Offset</label>
+            <Slider
+              min={-200}
+              max={200}
+              step={10}
+              value={[offset]}
+              onChange={({ value }) => setOffset(value[0])}
+              overrides={{
+                TickBar: {
+                  style: {
+                    display: "none"
+                  }
+                }
+              }}
+            />
+          </Block>
+
+          <Checkbox
+            checked={autoRefreshPlugin}
+            onChange={() => setAutoRefreshPlugin(!autoRefreshPlugin)}
```

Comment:
Do all of these onChanges need to return or just invoke?

Reply:
this as best practice as per Base Web components implementation  - onChange returns new value invoking react hook set function.